upb/decode.c

```diff
@@ -500,119 +565,168 @@ static const char *decode_tomsg(upb_decstate *d, const char *ptr, upb_msg *msg,
+      val.str_val.data = ptr;
+      val.str_val.size = size;
+      ptr += size;
+      op = delim_ops[ndx];
+      break;
+    }
+    case UPB_WIRE_TYPE_START_GROUP:
+      val.uint32_val = field_number;
+      op = OP_SUBMSG;
+      if (field->descriptortype != UPB_DTYPE_GROUP) goto unknown;
+      break;
+    case UPB_WIRE_TYPE_END_GROUP:
+      d->end_group = field_number;
+      ret.ptr = ptr;
+      ret.group_end = true;
+      return ret;
+    default:
+      decode_err(d);
+  }
+
+  if (op >= 0) {
+    /* Parse, using op for dispatch. */
+    switch (field->label) {
+      case UPB_LABEL_REPEATED:
+      case _UPB_LABEL_PACKED:
+        ptr = decode_toarray(d, ptr, msg, layout, field, val, op);
+        break;
+      case _UPB_LABEL_MAP:
+        decode_tomap(d, msg, layout, field, val);
         break;
-      case UPB_WIRE_TYPE_END_GROUP:
-        d->end_group = field_number;
-        return ptr;
       default:
-        decode_err(d);
+        ptr = decode_tomsg(d, ptr, msg, layout, field, val, op);
+        break;
     }
-
-    if (op >= 0) {
-      /* Parse, using op for dispatch. */
-      switch (field->label) {
-        case UPB_LABEL_REPEATED:
-        case _UPB_LABEL_PACKED:
-          ptr = decode_toarray(d, ptr, msg, layout, field, val, op);
-          break;
-        case _UPB_LABEL_MAP:
-          decode_tomap(d, msg, layout, field, val);
-          break;
-        default:
-          ptr = decode_tomsg(d, ptr, msg, layout, field, val, op);
-          break;
-      }
-    } else {
-    unknown:
-      /* Skip unknown field. */
-      if (field_number == 0) decode_err(d);
-      if (wire_type == UPB_WIRE_TYPE_START_GROUP) {
-        ptr = decode_group(d, ptr, NULL, NULL, field_number);
-      }
-      if (msg) {
-        if (!_upb_msg_addunknown(msg, field_start, ptr - field_start,
-                                 d->arena)) {
-          decode_err(d);
-        }
-      }
+  } else {
+  unknown:
+    /* Skip unknown field. */
+    if (field_number == 0) decode_err(d);
+    if (wire_type == UPB_WIRE_TYPE_START_GROUP) {
+      ptr = decode_group(d, ptr, NULL, NULL, field_number);
+    }
+    if (msg) {
+      decode_addunknown(d, msg, field_start, ptr - field_start);
     }
   }
 
+  ret.ptr = ptr;
+  ret.group_end = false;
+  return ret;
+}
+
+UPB_NOINLINE
+const char *fastdecode_generic(upb_decstate *d, const char *ptr, upb_msg *msg,
+                               const upb_msglayout *table, uint64_t hasbits,
+                               uint64_t data) {
+  decode_parseret ret;
+  *(uint32_t*)msg |= hasbits >> 16;  /* Sync hasbits. */
+  (void)data;
+  if (ptr == d->limit) return ptr;
+  ret = decode_field(d, ptr, msg, table);
+  if (ret.group_end) return ptr;
+  return fastdecode_dispatch(d, ret.ptr, msg, table, hasbits);
+}
+
+UPB_NOINLINE
+static const char *decode_msg(upb_decstate *d, const char *ptr, upb_msg *msg,
+                              const upb_msglayout *layout) {
+  if (msg) {
```

Comment:
how can this not always be true? I would expect layout

Reply:
Unknown fields. It's true that layout would be null there too.